src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -1122,15 +1198,45 @@ private void PerformSyncOperation<TOperation>(ref OperationQueue<TOperation> que
                     return;
                 }
 
-                if (e.Wait(timeout))
+                bool timeoutExpired = false;
+                while (true)
                 {
-                    // Completed within timeout
-                    return;
+                    DateTime waitStart = DateTime.UtcNow;
+
+                    if (!e.Wait(timeout))
+                    {
+                        timeoutExpired = true;
+                        break;
+                    }
+
+                    // Reset the event now to avoid lost notifications if the processing is unsuccessful.
+                    e.Reset();
+
+                    // We've been signalled to try to process the operation.
+                    OperationQueue<TOperation>.OperationResult result = queue.ProcessQueuedOperation(operation);
+                    if (result == OperationQueue<TOperation>.OperationResult.Completed ||
+                        result == OperationQueue<TOperation>.OperationResult.Cancelled)
+                    {
+                        break;
+                    }
+
+                    // Couldn't process the operation.
+                    // Adjust timeout and try again.
+                    if (timeout > 0)
+                    {
+                        timeout -= (DateTime.UtcNow - waitStart).Milliseconds;
+
+                        if (timeout <= 0)
+                        {
+                            timeoutExpired = true;
+                            break;
+                        }
+                    }
                 }
 
-                bool cancelled = operation.TryCancel();
-                if (cancelled)
+                if (timeoutExpired)
                 {
+                    queue.CancelAndContinueProcessing(operation);
                     operation.ErrorCode = SocketError.TimedOut;
                 }
             }
```

Comment:
Is there some test we can add to help flag this in the future?

Reply:
Flag what exactly?

Reply:
The case where we end up requiring two ThreadPool threads rather than one to handle a sync-over-async socket operation.

Reply:
I was trying to come up with a reasonable way to test this, but didn't come up with anything reasonable.  Suggestions?

Reply:
I'll send something this evening I think will work.

Reply:
I added a test here:
https://github.com/dotnet/corefx/pull/29117
It fails for me on Unix before this PR, and when I cherry-pick your changes, it then succeeds.

Reply:
Looks like the test is unhappy in CI...

Reply:
I just restarted the tests.  The failures I saw didn't seem to be this test, but maybe I missed them?

Reply:
Ah, ok

Reply:
I *think* they are just connectivity errors, so re-running.

Reply:
I'm on my phone right now and the mission control site is close to impossible to use on a mobile device :(